### ISSUE TITLE
`listitem` rule fails with roles inheriting from the 'list' or 'listitem' roles

### ISSUE THREAD

**Issue by rdeltour** [dequelabs/axe-core#1677]
Expectation: Axe would not report a failure to the rule `listitem` when native list elements (`ol`,`ul`,`li`) declare a more specific role _inheriting_ from their native role (e.g. [`role="directory"` on `ul`](https://www.w3.org/TR/wai-aria-1.1/#directory), [`role="doc-biblioentry` on `li`](https://www.w3.org/TR/dpub-aria-1.0/#doc-biblioentry), etc.).

Actual: the rule `listitem` fails.

Motivation: even if some of those roles may have limited accessibility support (like the DPUB roles), using them on elements with native list semantics is an appropriate fallback and should pass the rule.

<pre><code>
axe-core version: 3.2.2
</code></pre>

Example 1 of markup that will fail the rule `listitem`
```html
<ul>
  <li role="doc-biblioentry">an entry</li>
  …
</ul>
```

Example 2 of markup that will fail the rule `listitem`
```html
<ol role="directory">
   <li><a href="preface_001.xhtml">Original Transcriber’s Notes:</a></li>
   <li><a href="introduction_001.xhtml">ETYMOLOGY.</a></li>
   <li><a href="epigraph_001.xhtml">EXTRACTS (Supplied by a Sub-Sub-Librarian).</a></li>
   <li><a href="chapter_001.xhtml">Chapter 1. Loomings.</a></li>
   …
</ol>
```

**Comment by rdeltour** [comment]
If you agree this is a false-positive, I can submit at PR 😊

**Comment by straker** [comment]
Looks good to me, a PR would be great. Just to be clear, did you mean that 

```html
<ul>
  <li role="doc-biblioentry">an entry</li>
  …
</ul>
```

fails the `list` rule?

**Comment by WilcoFiers** [comment]
Totally agreed, although I'm not sure there's an obvious fix for it. Axe-core doesn't have a good representation of the ARIA role hierarchy. If you find it's going to be a complicated fix, please put together a proposal. Don't want to waste a whole bunch of time and then find out there's some (possibly unrelated) edge case that makes it not work.

**Comment by rdeltour** [comment]
@straker 
> did you mean that (…) fails the `list` rule?

Yes, that one would fail the `list` rule (`only-list-items` check), and the other example I gave in the OP would fail the `listitem` rule (`listitem` check).

@WilcoFiers 
>  I'm not sure there's an obvious fix for it. Axe-core doesn't have a good representation of the ARIA role hierarchy.

In our local patches to Axe we used the `axe.commons.aria.getRoleType(role)` API, which is based on the lookup table. It may not be perfect (i.e. will not support transitive inheritance) but should be enough in this case?
See [our patch for the `only-list-items` check](https://github.com/daisy/ace/blob/v1.1.1/packages/ace-core/src/scripts/axe-patch-only-list-items.js#L16) and [our patch for the `listitem` check](https://github.com/daisy/ace/blob/v1.1.1/packages/ace-core/src/scripts/axe-patch-listitem.js#L27).

**Comment by straker** [comment]
There are other role types that aren't `listitem` that can be used on `<li>` items that will still fail the `list` rule (e.g. `menuitemradio`, `menuitemcheckbox`). Would it be possible to use the `aria.isAriaRoleAllowedOnElement` api to see if the role is allowed on the element instead of just looking at list roles?

**Comment by WilcoFiers** [comment]
@rdeltour I was just reviewing this ACT-R rules:
https://github.com/act-rules/act-rules.github.io/pull/255/files

I started digging a little, and according to dpub-aria, `doc-biblioentry` has `doc-bibliography` as a required context, so a `ul` without `role="doc-bibliography"` should fail, no?

**Comment by rdeltour** [comment]
> I started digging a little, and according to dpub-aria, `doc-biblioentry` has `doc-bibliography` as a required context, so a `ul` without `role="doc-bibliography"` should fail, no?

Yes, I think it should fail.

**Comment by WilcoFiers** [comment]
Looking at this again. At this point both `directory` and `doc-biblioentry` are proposed to be deprecated, exactly for the reason that its inheritance creates confusion. Neither of these use cases is valid according to ARIA 1.1, so we're going to leave these as failures for axe-core.
